### PR TITLE
Wandb with Ray Integration

### PR DESCRIPTION
**Pull request by mvacaporale**
The following PR includes a Wandb integration with ray. Changes include
* `ray_wandb.py` with `WandbLogger` class
* Added ray serializer and deserializer for torch tensors - needed for logging array-like values without errors.

### CODE REVIEW

**Comment by mvacaporale** [comment]
The `WandbLogger` is still in progress and probably won't be complete until the Wandb fixes some issues with distributed training. Nonetheless, I'd appreciate feedback on what I've been using so far.

@lscheinkman I'm curious if you have any thoughts on the de/serializers added to `imagenet_tune.py`
@lucasosouza If you're interested in using the integration see `ray_wandb.WandbLogger` - The other functions in that file are still in progress.

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py [hunk 1]
```diff
@@ -341,6 +341,33 @@ def run(config):
     address = os.environ.get("REDIS_ADDRESS", config.get("redis_address"))
     ray.init(address=address, local_mode=config.get("local_mode", False))
 
+    # Register serializer and deserializer - needed when logging arrays and tensors.
+    def serializer(obj):
+        if obj.requires_grad:
+            obj = obj.detach()
+        if obj.is_cuda:
+            return obj.cpu().numpy()
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:349]: How often and when is this called? 
Coping tensor from GPU to CPU could slow everything down If it is called inside the training or evaluation loop

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py [hunk 2]
```diff
@@ -341,6 +341,33 @@ def run(config):
...
+        torch.DoubleTensor,
+        torch.HalfTensor,
+        torch.ByteTensor,
+        torch.CharTensor,
+        torch.ShortTensor,
+        torch.IntTensor,
+        torch.LongTensor,
+        torch.Tensor,
+    ]:
+        def deserializer(serialized_obj):
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:364]: Dont you need to know the destination device (cpu or gpu) when you deserialize?

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py [hunk 3]
```diff
@@ -341,6 +341,33 @@ def run(config):
...
+            return tensor_type(serialized_obj)  # cast to tensor_type
+
+        ray.register_custom_serializer(
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:367]: Due to the extra GPU/CPU transfer involved maybe you make this optional, only register if using `wandb`

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 1 above)
Comment (mrcslws) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:349]: In case it's useful, I've found it best to accumulate logs on the GPU and then copy them to the CPU at the end of the epoch. If you do it that way, there is essentially zero performance cost. Here's example code that I'll submit as a PR soon, after my other PR is merged: https://github.com/mrcslws/nupic.research/blob/907cf9709cc5fd741326baf23a7de5893c7cbcfd/nupic/research/frameworks/pytorch/imagenet/mixins/log_every_loss.py#L65

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 3 above)
Comment (mrcslws) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:367]: It seems reasonable to register these by default? This just makes it so that we can log tensors, include them in the result dict, etc. Are you thinking people should get an error when they try to do this, so that they explicitly account for performance?

I'd maybe put it in a separate function so that I can call it if I'm not using `run`.

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 1 above)
Comment (mrcslws) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:349]: (In case it's not clear, this comment^ is not a suggestion for this PR. It's a guideline for how we can responsibly add more tracing to our code, whether it's via wandb or via the ray result dict.)

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py [hunk 4]
```diff
@@ -341,6 +341,33 @@ def run(config):
     address = os.environ.get("REDIS_ADDRESS", config.get("redis_address"))
     ray.init(address=address, local_mode=config.get("local_mode", False))
 
+    # Register serializer and deserializer - needed when logging arrays and tensors.
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:344]: The docs say:
The serializer and deserializer are used when transferring objects of `cls` across processes and nodes. This can be significantly faster than the Ray default fallbacks. Wraps `register_custom_serializer` underneath.

It's thus not used exclusively by the loggers, but it is invoked when the logs have array-like objects. In practice, I've only seen them used on the logs though.

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 1 above)
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:349]: As mentioned in the last comment. I've only seen this called on the logs. I also only log cpu values, but the `is_cuda` is included just in case.

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 2 above)
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:364]: That's a good point. So far, I haven't seen this cause any issues, but that may be due to the limited application I've testing (with the logs).

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:344]: Did you test with distributed training using multiple GPU nodes?

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 3 above)
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:367]: As for registering it only if using wandb: It's not just for that purpose. It's for all logging. I include it with this PR because I came across its benefit as I was testing the wandb integration.

That said, I can add it to a separate function for other use cases.

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:344]: I did. That was the main case that inspired the addition. What's your line of thinking?

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 3 above)
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:370]: I added the de/serializers to a separate function `register_torch_serializers` in `nupic.research.support.ray_utils.py`.

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:347]: I am concerned about `obj.cpu().numpy()` call. We just need to make sure we understand the performance cost of adding this extra serialization. If you tested with multiple nodes and the serialization was only used by the loggers and you did not notice any performance loss then it should be fine

[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:347]: That's a fair concern. I can attempt a test today to double check. I'm 90% sure, but you're right in the care needed to keep performance up.
